File: 22.Automated-Testing-Unit-Integration-and-E2E/code/python/anatomy_of_a_test_1.py
# pytest discovers files named test_*.py and functions named test_*.
# A bare `assert` is enough ,  pytest rewrites it to show a rich failure diff.
def test_discount_applies_percentage():
    # Arrange ,  set up inputs (and any doubles)
    cart = Cart(subtotal=200)
    coupon = Coupon(percent=10)

    # Act ,  the ONE operation under test
    total = apply_discount(cart, coupon)

    # Assert ,  pytest prints actual vs expected automatically on failure
    assert total == 180

File: 22.Automated-Testing-Unit-Integration-and-E2E/code/go/anatomy_of_a_test_1.go
// File naming convention: foo.go is tested by foo_test.go in the same package.
// Test functions are func TestXxx(t *testing.T) ,  the `go test` tool finds them.
func TestDiscount_AppliesPercentage(t *testing.T) {
    // Arrange ,  set up inputs (and any doubles)
    cart := Cart{Subtotal: 200}
    coupon := Coupon{Percent: 10}

    // Act ,  the ONE operation under test
    total := ApplyDiscount(cart, coupon)

    // Assert ,  state the expectation; a good message explains the failure
    if total != 180 {
        t.Errorf("ApplyDiscount(200, 10%%) = %d; want 180", total)
    }
}
